### PR TITLE
Replace SWIG bindings with Pybind11, and remove SWIG dependency

### LINKED ISSUE #2646: Create SWIG + Deprecation Unit Example Demonstrating Syntax Error
In #2597, marking a method as deprecated resulted in SWIG complaining about a syntax error. The current work around is to add an extra set of `#ifndef SWIG .... #endif` guards to ensure the `DRAKE_DEPRECATED()` macro is not included when SWIG is in use. We should create a unit example  demonstrating that this problem continues to exist in isolation.

See: https://reviewable.io/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd

### LINKED ISSUE #4243: Mysterious SWIG-related error
@jamiesnape reports that it has happened before. Two of the more recent incidents:

https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=214150
https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=214070

Error message is not useful, unsure of who exactly to assign this to, which is why I did software core.

### LINKED ISSUE #4454: rbtree.py Is Not Automatically Regenerated When A New Header Is Included In rbtree.i
# Problem Definition

When a new header file is included in `drake-distro/drake/bindings/swig/rbtree.i`, the classes that are defined within this header file do not appear in the automatically-generated `drake-distro/drake/bindings/python/pydrake/rbtree.py` even after rebuilding `drake`. Oddly, the new classes' symbols _do_ appear in `drake-distro/build/drake/bindings/swig/_rbtree.so`, meaning the two are inconsistent. Deleting `drake-distro/drake/bindings/python/pydrake/rbtree.py` and rebuilding `drake` does not work because `ninja` does not believe anything changed (I'm currently assuming `make` will behave identically).

# Workaround

The only workaround I've found is to do a comprehensive nuke:

```
$ cd drake-distro
$ rm -rf *
$ git reset --hard HEAD
$ mkdir build; cd build
$ cmake .. -G Ninja
$ ninja
```

I'm wondering whether there's a better way to regenerate the correct `rbtree.py` without resorting to the comprehensive nuke.

### LINKED ISSUE #4469: Warning - SWIG / Python Detected a Memory Leak
# Problem Definition

The following warning occurs when the SWIG / Python unit test `test_python_pydrake.test.testPR2IK` or `test_python_pydrake.test.test_urdf_parser` are run:

```
swig/python detected a memory leak of type 'std::map< std::string,int,std::less< std::string >,std::allocator< std::pair< std::string const,int > > > *', no destructor found.
```

# Examples:
* https://drake-cdash.csail.mit.edu/testDetails.php?test=5480982&build=251903
* https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-clang-continuous-debug/1387/

### LINKED ISSUE #4486: Reconsider the use of SWIG
# Problem Definition

Numerous oddities involving SWIG have cropped up:

* #4469 
* #4454 
* #4243 
* #2646

# Action Item

We should reconsider whether to continue using SWIG or switch to something else that is perhaps more target language specific. For example, for Python bindings, @jwnimmer-tri recommended [Boost::Python](http://www.boost.org/doc/libs/1_62_0/libs/python/doc/html/index.html) and @jamiesnape recommended [pybind11](https://github.com/pybind/pybind11).

### PR DESCRIPTION
**Pull request by rdeits**
~~This isn't quite ready for review yet, but I want to see how it does on CI. This also includes the changes from #4944 which should come in first. I'll update this branch when that happens.~~

This replaces all of the `pydrake` bindings with identical bindings generated with Pybind11. With this change and #4944 we no longer have any dependencies on SWIG, so this also deletes the SWIG and swigmake externals. 

This also:

* changes the python binding build process, so there should no longer be any python build files in the source tree
* removes all the `#ifndef SWIG` guards

Resolves #4486 
Fixes #4454 
Fixes #4243 
Fixes #2646 
Fixes #4469 

Todo before merging: 

- [x] Undo my evil hacks to the build system
- [x] Make sure there aren't any other methods that SWIG was generating that aren't present here
- [x] Discuss cleaning up the Python APIs

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/4949)

### CODE REVIEW

File: cmake/config.cmake [hunk 1]
```diff
@@ -204,11 +204,12 @@ macro(drake_setup_python)
...
-    find_package(Python 3 MODULE REQUIRED)
-  else()
-    find_package(Python 2.7 MODULE REQUIRED)
-  endif()
+  # if(WITH_PYTHON_3)
+  #   find_package(Python 3 MODULE REQUIRED)
+  # else()
+  #   find_package(Python 2.7 MODULE REQUIRED)
+  # endif()
+  find_package(pybind11 REQUIRED)
```
Comment (rdeits) [inline cmake/config.cmake]: Todo: something that isn't this

File: drake/CMakeLists.txt [hunk 2]
```diff
@@ -61,7 +61,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror=all -Werror=ignored-qualifiers -
 if(NOT CMAKE_CXX_FLAGS MATCHES "-fsanitize=") # sanitizers are extremely brittle without lazy linking
   if(APPLE)
     set(CMAKE_SHARED_LINKER_FLAGS "-Wl,-undefined -Wl,error ${CMAKE_SHARED_LINKER_FLAGS}")
-    set(CMAKE_MODULE_LINKER_FLAGS "-Wl,-undefined -Wl,error ${CMAKE_MODULE_LINKER_FLAGS}")
+    # set(CMAKE_MODULE_LINKER_FLAGS "-Wl,-undefined -Wl,error ${CMAKE_MODULE_LINKER_FLAGS}")
```
Comment (rdeits) [inline drake/CMakeLists.txt]: Todo: restore this flag in a way that doesn't break the python bindings

**Comment by rdeits** [comment]
Ok, modulo the clang issue, this is getting pretty close to ready. @patmarion what do I need to do to make sure all the director IK interfaces are working?

**Comment by rdeits** [comment]
More cmake issues here... I'm currently running into the fact that pybind11 ships its own `FindPythonLibsNew.cmake`, but that file won't be run if python has already been found. Not running it causes various subtle bugs like the python modules having the wrong file extension. I'm currently hacking around it by just calling `find_package(pybind11)` earlier in the drake build process, but that's not a good solution.

**Comment by rdeits** [comment]
@patmarion OK, pydrakeik tests are now working.

**Comment by rdeits** [comment]
Ok, this is good enough that I'd appreciate some feedback on it, and some help resolving the clang issues.

**Comment by rdeits** [comment]
The clang issues appear to be due to https://github.com/pybind/pybind11/issues/639

**Comment by rdeits** [comment]
And that issue is fixed in https://github.com/pybind/pybind11/pull/604

**Comment by rdeits** [comment]
Huge thanks to @mwoehlke-kitware for figuring out the automagic Eigen-Numpy conversions for custom scalar types. That's let me substantially simplify the python wrapper code, and will also help with #4771 . I've also switched over to our temporary pybind11 fork, which should (hopefully) resolve the clang issues.

**Comment by rdeits** [comment]
Review status: 0 of 47 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/CMakeLists.txt, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KbYkLL1IkeU8dW5vteI-r1-64:-KcURoiAX5I2G9M1r0tG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37bcec3e8d0f1ee1e6f5bb0e618b03f79d303ec0/drake/CMakeLists.txt#L64)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary><blockquote>

Todo: restore this flag in a way that doesn't break the python bindings
</blockquote></details>

Done.

---

*[cmake/config.cmake, line 212 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KbYkLL1IkeU8dW5vteG-r1-212:-KcURpHOHAUuNAqWOZTJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37bcec3e8d0f1ee1e6f5bb0e618b03f79d303ec0/cmake/config.cmake#L212)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary><blockquote>

Todo: something that isn't this
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
woohoo! this is finally working!

**Comment by rdeits** [comment]
+@sammy-tri  +@jamiesnape +@mwoehlke-kitware 

---

Review status: 0 of 47 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-:-KcUW3_OcvEQH6SK2UG4:bk8awr8)*

**Comment by jamiesnape** [comment]
Reviewed 21 of 47 files at r1, 1 of 2 files at r2, 6 of 27 files at r3, 5 of 13 files at r5, 2 of 10 files at r6.
Review status: 34 of 47 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/pybind11/CMakeLists.txt, line 9 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUX6G1AHa5Vl9n5LNa:-KcUX6G1AHa5Vl9n5LNb:b-9faue5) ([raw file](https://github.com/robotlocomotion/drake/blob/cfdf3f4a3560fc1b19be68393583383732812d3e/drake/bindings/pybind11/CMakeLists.txt#L9)):*
> ```CMake
> endif()
> 
> # Look up the python module suffix (which varies by python version).
> ```

Could this be added to [FindPython.cmake](https://github.com/RobotLocomotion/drake/blob/master/cmake/modules/FindPython.cmake)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
Review status: 34 of 47 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/pybind11/CMakeLists.txt, line 9 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUX6G1AHa5Vl9n5LNa:-KcUbVo29JjMIezmYkMT:b-q1p6nu) ([raw file](https://github.com/robotlocomotion/drake/blob/cfdf3f4a3560fc1b19be68393583383732812d3e/drake/bindings/pybind11/CMakeLists.txt#L9)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Could this be added to [FindPython.cmake](https://github.com/RobotLocomotion/drake/blob/master/cmake/modules/FindPython.cmake)?
</blockquote></details>

good point. I forgot we were shipping our own FindPython

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
Review status: 34 of 48 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/pybind11/CMakeLists.txt, line 9 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUX6G1AHa5Vl9n5LNa:-KcUe8Is7wBOzlKxzdjK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cfdf3f4a3560fc1b19be68393583383732812d3e/drake/bindings/pybind11/CMakeLists.txt#L9)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary><blockquote>

good point. I forgot we were shipping our own FindPython
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by jamiesnape** [comment]
Review status: 34 of 48 files reviewed at latest revision, 2 unresolved discussions.

---

*[cmake/modules/FindPython.cmake, line 38 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUf0qMwjzYQojYSS1d:-KcUf0qMwjzYQojYSS1e:b-80590f) ([raw file](https://github.com/robotlocomotion/drake/blob/cb2affe0e82393ed018ef6cab661c0fcda1e0ff8/cmake/modules/FindPython.cmake#L38)):*
> ```CMake
>     OUTPUT_STRIP_TRAILING_WHITESPACE)
> 
> if(NOT PYTHON_MODULE_EXTENSION_SUCCESS MATCHES 0)
> ```

Remove this and add PYTHON_MODULE_EXTENSION to REQUIRED_VARS. If PYTHON_MODULE_EXTENSION is empty, then find_package will fail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
Review status: 34 of 48 files reviewed at latest revision, 2 unresolved discussions.

---

*[cmake/modules/FindPython.cmake, line 38 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUf0qMwjzYQojYSS1d:-KcUfirdB36cA5cLV1FN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cb2affe0e82393ed018ef6cab661c0fcda1e0ff8/cmake/modules/FindPython.cmake#L38)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Remove this and add PYTHON_MODULE_EXTENSION to REQUIRED_VARS. If PYTHON_MODULE_EXTENSION is empty, then find_package will fail.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by jamiesnape** [comment]
Review status: 34 of 48 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cmake/modules/FindPython.cmake, line 38 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUf0qMwjzYQojYSS1d:-KcUg9YqYb0ZpkajPWdh:bvbqi8j) ([raw file](https://github.com/robotlocomotion/drake/blob/cb2affe0e82393ed018ef6cab661c0fcda1e0ff8/cmake/modules/FindPython.cmake#L38)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

You should only set `PYTHON_MODULE_EXTENSION` if `PYTHON_MODULE_EXTENSION_SUCCESS` indicates success. You probably do not need to capture `ERROR_VARIABLE`.

---

*[cmake/modules/FindPython.cmake, line 40 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUft4TFF7Aip1ms3E6:-KcUft4TFF7Aip1ms3E7:b-xvksbv) ([raw file](https://github.com/robotlocomotion/drake/blob/dcc8428c98f30c334efd5c9d5daed629e8607fe4/cmake/modules/FindPython.cmake#L40)):*
> ```CMake
> find_package_handle_standard_args(Python
>   REQUIRED_VARS PYTHON_EXECUTABLE PYTHON_INCLUDE_DIR PYTHON_LIBRARY
>   VERSION_VAR PYTHON_VERSION_STRING PYTHON_MODULE_EXTENSION)
> ```

Needs to be on the line above. `VERSION_VAR` is the name of an argument rather than a required variable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
Review status: 34 of 48 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/modules/FindPython.cmake, line 38 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUf0qMwjzYQojYSS1d:-KcUl6VNRg7Np_w8vKFZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cb2affe0e82393ed018ef6cab661c0fcda1e0ff8/cmake/modules/FindPython.cmake#L38)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

You should only set `PYTHON_MODULE_EXTENSION` if `PYTHON_MODULE_EXTENSION_SUCCESS` indicates success. You probably do not need to capture `ERROR_VARIABLE`.
</blockquote></details>

Done.

---

*[cmake/modules/FindPython.cmake, line 40 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcUft4TFF7Aip1ms3E6:-KcUkWBhvgmp9QW-spDA:b5hp82a) ([raw file](https://github.com/robotlocomotion/drake/blob/dcc8428c98f30c334efd5c9d5daed629e8607fe4/cmake/modules/FindPython.cmake#L40)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Needs to be on the line above. `VERSION_VAR` is the name of an argument rather than a required variable.
</blockquote></details>

Agh, sorry. Done. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by sammy-tri** [comment]
+(status: curate commits before merging) 

Partial pass completed.  Left a few comments.

---

Reviewed 21 of 47 files at r1, 10 of 27 files at r3, 1 of 9 files at r4, 3 of 13 files at r5, 7 of 10 files at r6, 1 of 1 files at r7, 1 of 2 files at r8, 1 of 1 files at r10.
Review status: 46 of 48 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 1 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYiO7m7m8T4IDcZHhz:-KcYiO7nfF5Zn53IiK5o:bd0dj9f) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/CMakeLists.txt#L1)):*
> ```CMake
> if(NOT DISABLE_PYTHON)
> ```

BTW It seems odd to duplicate this check in drake/CMakeLists.txt and this file.

---

*[drake/bindings/pybind11/pydrake_autodiffutils.cc, line 13 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYjI7hU2aX1w4SZCkU:-KcYjI7isRhOiJp11HZE:b-l0dcfe) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/pybind11/pydrake_autodiffutils.cc#L13)):*
> ```cpp
> using std::cos;
> 
> template <typename Derived>
> ```

BTW this function could maybe use a comment...  it's not obvious where an AutoDiffScalar templated on a Derived which isn't VectorXd would come from.

---

*[drake/bindings/pybind11/pydrake_rbtree.cc, line 38 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYk1MySNTJO9N6fHwC:-KcYk1MySNTJO9N6fHwD:b3k9qc6) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/pybind11/pydrake_rbtree.cc#L38)):*
> ```cpp
>                 urdf_filename, floating_base_type, &instance);
>         },
>         py::arg("urdf_filename"), py::arg("joint_type")="ROLLPITCHYAW"
> ```

Missing whitespace around = operator

---

*[drake/bindings/python/pydrake/__init__.py, line 1 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYlJNSe6YBEVItyfRR:-KcYlJNTA-UO_rq5a88p:b-ua0n0e) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/python/pydrake/__init__.py#L1)):*
> ```Python
> from __future__ import absolute_import, division, print_function
> ```

do division and print_function do anything by being imported in this file?  (or several other places where they're used?)

---

*[drake/bindings/python/pydrake/autodiffutils.py, line 3 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYlTsalMqVFrHAoyel:-KcYlTsbAcPOUDjLznII:bsrw31x) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/python/pydrake/autodiffutils.py#L3)):*
> ```Python
> from __future__ import absolute_import, division, print_function
> 
> import numpy as np
> ```

What effect does this have here when we don't reference np later?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-:-KcYhvJ-Bt8qVlWizv7r:b-1n9mwm)*

**Comment by rdeits** [comment]
This currently doesn't compile on OSX due to issues finding the numpy headers. We may need one more update to our pybind11 fork

**Comment by patmarion** [comment]
btw, director uses a copy of this file to find NumPy headers:

https://raw.githubusercontent.com/pydata/numexpr/master/FindNumPy.cmake

and uses it like this:

```
find_package(NumPy)

if (NOT NUMPY_FOUND)
  message(WARNING "Optimized transformations C extension module is disabled because numpy is not found.")
  return()
endif()

if (NUMPY_VERSION VERSION_LESS 1.7)
  message(WARNING "Optimized transformations C extension module is disabled because numpy version < 1.7 was found.")
  return()
endif()


include_directories(${NUMPY_INCLUDE_DIRS})
```

**Comment by rdeits** [comment]
Review status: 42 of 48 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 1 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYiO7m7m8T4IDcZHhz:-KcYpU44Q3riLByxMpmz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/CMakeLists.txt#L1)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW It seems odd to duplicate this check in drake/CMakeLists.txt and this file.
</blockquote></details>

Done.

---

*[drake/bindings/pybind11/pydrake_autodiffutils.cc, line 13 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYjI7hU2aX1w4SZCkU:-KcYpV1JGImmGNyaVTTa:b-jpsucd) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/pybind11/pydrake_autodiffutils.cc#L13)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW this function could maybe use a comment...  it's not obvious where an AutoDiffScalar templated on a Derived which isn't VectorXd would come from.
</blockquote></details>

Done. The other Derived type is `Eigen::CWiseBinaryOp` which represents an un-evaluated computation. 

---

*[drake/bindings/pybind11/pydrake_rbtree.cc, line 38 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYk1MySNTJO9N6fHwC:-KcYpcD-d8PeY9EcO_xi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/pybind11/pydrake_rbtree.cc#L38)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Missing whitespace around = operator
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/__init__.py, line 1 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYlJNSe6YBEVItyfRR:-KcYoVZdx_SRlw3paXrb:b47c2z5) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/python/pydrake/__init__.py#L1)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

do division and print_function do anything by being imported in this file?  (or several other places where they're used?)
</blockquote></details>

No, they have no effect. I'm personally trying t get into the habit of just adding this boilerplate to the top of all my python files because I'm accustomed to the python3 behavior of division and `print()` and would like to use them everywhere. 

This is different from an unused import of a package because it doesn't actually load any code, and it prevents subtle bugs later on. 

For example, if I forget to import numpy and then try to write `numpy.zeros()`, I get a helpful error message that numpy was not found. But if I forget to `from __future__ import division` and then try to write `1 / 2`, I don't get an error: I just get the wrong answer. So, better to just have the import and not need it than to not have it and cause problems later. 

---

*[drake/bindings/python/pydrake/autodiffutils.py, line 3 at r10](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcYlTsalMqVFrHAoyel:-KcYpg7fn3JKXE4K3tk3:b-cwokoy) ([raw file](https://github.com/robotlocomotion/drake/blob/38c4cf5d420f0d09eaf1531a06e63b88fecd4411/drake/bindings/python/pydrake/autodiffutils.py#L3)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

What effect does this have here when we don't reference np later?
</blockquote></details>

Done. Just a leftover import that I forgot to remove. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
@patmarion ok, I've added that cmake script to pybind11 and updated the drake cmakelists. I'll see if that gets everything working again.

**Comment by rdeits** [comment]
Ok, I can now build this on OSX. @mwoehlke-kitware do you think including the FindNumpy.cmake in pybind11 is a good solution? Or can we avoid including that header in pybind11 and thus avoid the dependency?

**Comment by sammy-tri** [comment]
:lgtm: pending open discussions

---

Reviewed 1 of 27 files at r3, 4 of 13 files at r5, 2 of 10 files at r6, 8 of 8 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/python/pydrake/rbtree.py, line 3 at r11](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcZBXrAbvbd10K4cdXb:-KcZBXrAbvbd10K4cdXc:b-eobs2x) ([raw file](https://github.com/robotlocomotion/drake/blob/61b2b4420275ca9e1d1b622cf3cf804d8507c22d/drake/bindings/python/pydrake/rbtree.py#L3)):*
> ```Python
> from __future__ import absolute_import, division, print_function
> 
> import numpy as np
> ```

Another leftover import?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-:-KcZBrX3E_CHGmqPXSB9:blm8kzx)*

**Comment by rdeits** [comment]
Review status: 48 of 49 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/python/pydrake/rbtree.py, line 3 at r11](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcZBXrAbvbd10K4cdXb:-KcZHF8B8jwOklAGyLsQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61b2b4420275ca9e1d1b622cf3cf804d8507c22d/drake/bindings/python/pydrake/rbtree.py#L3)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Another leftover import?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
@drake-jenkins-bot retest this please

**Comment by jamiesnape** [comment]
:lgtm:

---

Reviewed 4 of 27 files at r3, 1 of 9 files at r4, 2 of 13 files at r5, 3 of 10 files at r6, 1 of 1 files at r7, 1 of 1 files at r10, 8 of 8 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-:-KcZcdv5ywlGlzwTHpuS:bnfp4nl)*

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-bazel-experimental please

**Comment by rdeits** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by mwoehlke-kitware** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/bindings/pybind11/pydrake_rbtree.cc, line 25 at r12](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcceaoiQSh252sM9SKP:-KcceaolQXsZekuwJNr5:b-5ikkkz) ([raw file](https://github.com/robotlocomotion/drake/blob/da16d5f27addcd72200c98976b9626484912838c/drake/bindings/pybind11/pydrake_rbtree.cc#L25)):*
> ```cpp
>             drake::multibody::joints::FloatingBaseType floating_base_type;
> 
>             if (joint_type == "FIXED") {
> ```

BTW, is there a reason we don't wrap this enum _as_ an enum?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by mwoehlke-kitware** [comment]
> do you think including the FindNumpy.cmake in pybind11 is a good solution? Or can we avoid including that header in pybind11 and thus avoid the dependency?

The only way I see to do that is to build a temporary Python list (or list of lists, for the 2D case) and convert that to a numpy array instead of building it directly. Unfortunately, the NumPy "function" to set an item in an array is not accessible via the "capsule" API that can be used without the NumPy headers.

It looks like the "right" solution would be a) including `FindNumpy.cmake` in pybind11, and b) calling it via `pybind11Config.cmake`. The current work-around has essentially the same effect; we could leave it for now and fix it later (we'll need to before this stuff can be merged upstream), if that's okay with everyone?

(Pedantic: shouldn't that file be `FindNumPy.cmake`, with a `P`, not a `p`?)

**Comment by rdeits** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/bindings/pybind11/pydrake_rbtree.cc, line 25 at r12](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcceaoiQSh252sM9SKP:-Kcchc5ldzxzDt-iqWiM:b-saxzbq) ([raw file](https://github.com/robotlocomotion/drake/blob/da16d5f27addcd72200c98976b9626484912838c/drake/bindings/pybind11/pydrake_rbtree.cc#L25)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

BTW, is there a reason we don't wrap this enum _as_ an enum?
</blockquote></details>

No, there isn't. In fact, we *do* wrap that enum as an enum elsewhere. I kept the string argument because it's what the existing python bindings do. I'll add fixing this to the Python API todos. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by mwoehlke-kitware** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/bindings/pybind11/pydrake_rbtree.cc, line 25 at r12](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-KcceaoiQSh252sM9SKP:-KcczoCPrWKfhLr0LZMH:b-5znt9v) ([raw file](https://github.com/robotlocomotion/drake/blob/da16d5f27addcd72200c98976b9626484912838c/drake/bindings/pybind11/pydrake_rbtree.cc#L25)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary><blockquote>

No, there isn't. In fact, we *do* wrap that enum as an enum elsewhere. I kept the string argument because it's what the existing python bindings do. I'll add fixing this to the Python API todos. 
</blockquote></details>

Okay, that seems like a good plan. (I suspected the reason was "compatibility", but yeah, it would be nice if we can fix it at some point... The string seems less discoverable and slightly more error-prone.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
The updates to pybind11 caused issues with IRIS, which I've (hopefully) addressed in https://github.com/rdeits/iris-distro/pull/55

**Comment by rdeits** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by rdeits** [comment]
Oops, had a conflict with master. Let's try this again, post-rebase:

@drake-jenkins-bot mac-clang-experimental please

**Comment by rdeits** [comment]
-@mwoehlke-kitware since this already has two assigned reviewers

---

Review status: 41 of 50 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-:-Kcf_hDmV6c7gLTUEjEO:b-yk9c22)*

**Comment by rdeits** [comment]
I don't know what's up with the `linux-gcc-experimental-ros` build. Is that a new one?

**Comment by jwnimmer-tri** [comment]
Its another #3429 flake.
@drake-jenkins-bot linux-gcc-experimental-ros please

**Comment by jamiesnape** [comment]
:lgtm:

---

Reviewed 8 of 8 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949#-:-Kcrq45zs7XmRghU_VJQ:bnfp4nl)*

**Comment by sammy-tri** [comment]
Reviewed 1 of 1 files at r12, 8 of 8 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4949)*

**Comment by rdeits** [comment]
Is this good to go? Reviewable doesn't like my commit history, but I assume we can just squash-merge this PR in, right?

**Comment by jamiesnape** [comment]
Yes to both questions.

**Comment by SeanCurtis-TRI** [comment]
For the record, does this produce a bunch of files that should be included in .gitignore?

I pulled master and built, and now I've got a number of files in `drake/bindings/python` as untracked changes.

**Comment by jamiesnape** [comment]
Should be the opposite. You have old SWIG build products.

**Comment by SeanCurtis-TRI** [comment]
Ah...so you're suggesting they didn't appear as part of the build process, but in simply updating my master.  And, in fact, these are files that are old and are just now visible to git.  I certainly have `externals/swigmake` and `externals/swig_matlab` appearing.  So, I guess the `drake/bindings/python/pydrake/` stuff is related.  Thanks.

**Comment by rdeits** [comment]
Yeah, I removed the swig build products from .gitignore, so you're only
seeing them now.

**Comment by EricCousineau-TRI** [comment]
Linking to #1267.

**Pull request merged**